pvlib/pvsystem.py

```diff
@@ -2137,62 +2296,78 @@ def i_from_v(resistance_shunt, resistance_series, nNsVth, voltage,
-        # Recast in terms of Gsh=1/Rsh for better numerical stability.
-        I[idx_p] = (IL[idx_p] + I0[idx_p] - V[idx_p]*Gsh[idx_p]) / \
-            (Rs[idx_p]*Gsh[idx_p] + 1.) - (a[idx_p]/Rs[idx_p])*lambertwterm
-
-    if output_is_scalar:
-        return np.asscalar(I)
+    if method.lower() == 'lambertw':
+        try:
+            from scipy.special import lambertw
+        except ImportError:
+            raise ImportError('This function requires scipy')
+
+        # Record if inputs were all scalar
+        output_is_scalar = all(map(np.isscalar,
+                                   [resistance_shunt, resistance_series, nNsVth,
+                                    voltage, saturation_current, photocurrent]))
+
+        # This transforms Gsh=1/Rsh, including ideal Rsh=np.inf into Gsh=0., which
+        #  is generally more numerically stable
+        conductance_shunt = 1./resistance_shunt
+
+        # Ensure that we are working with read-only views of numpy arrays
+        # Turns Series into arrays so that we don't have to worry about
+        #  multidimensional broadcasting failing
+        Gsh, Rs, a, V, I0, IL = \
+            np.broadcast_arrays(conductance_shunt, resistance_series, nNsVth,
+                                voltage, saturation_current, photocurrent)
+
+        # Intitalize output I (V might not be float64)
+        I = np.full_like(V, np.nan, dtype=np.float64)
+
+        # Determine indices where 0 < Rs requires implicit model solution
+        idx_p = 0. < Rs
+
+        # Determine indices where 0 = Rs allows explicit model solution
+        idx_z = 0. == Rs
+
+        # Explicit solutions where Rs=0
+        if np.any(idx_z):
+            I[idx_z] = IL[idx_z] - I0[idx_z]*np.expm1(V[idx_z]/a[idx_z]) - \
+                Gsh[idx_z]*V[idx_z]
+
+        # Only compute using LambertW if there are cases with Rs>0
+        # Does NOT handle possibility of overflow, github issue 298
+        if np.any(idx_p):
+            # LambertW argument, cannot be float128, may overflow to np.inf
+            argW = Rs[idx_p]*I0[idx_p]/(a[idx_p]*(Rs[idx_p]*Gsh[idx_p] + 1.)) * \
+                np.exp((Rs[idx_p]*(IL[idx_p] + I0[idx_p]) + V[idx_p]) /
+                       (a[idx_p]*(Rs[idx_p]*Gsh[idx_p] + 1.)))
+
+            # lambertw typically returns complex value with zero imaginary part
+            # may overflow to np.inf
+            lambertwterm = lambertw(argW).real
+
+            # Eqn. 2 in Jain and Kapoor, 2004
+            #  I = -V/(Rs + Rsh) - (a/Rs)*lambertwterm + Rsh*(IL + I0)/(Rs + Rsh)
+            # Recast in terms of Gsh=1/Rsh for better numerical stability.
+            I[idx_p] = (IL[idx_p] + I0[idx_p] - V[idx_p]*Gsh[idx_p]) / \
+                (Rs[idx_p]*Gsh[idx_p] + 1.) - (a[idx_p]/Rs[idx_p])*lambertwterm
+
+        if output_is_scalar:
+            return np.asscalar(I)
+        else:
+            return I
     else:
+        # use singlediode methods
```

Comment:
Replace with # Calculate points on the IV curve using either 'newton' or 'brentq' methods. Voltages are determined by first solving the single diode equation for the diode voltage V_d then backing out voltage

Reply:
✔️